### PR TITLE
define telemetry protocol

### LINKED ISSUE #4: Create proto file for telemetry plugin

### PR DESCRIPTION
**Pull request by JonasVautherin**
Fixes #4.

### CODE REVIEW

File: telemetry/telemetry.proto [hunk 1]
```diff
@@ -6,15 +6,90 @@ option java_package = "io.dronecore.telemetry";
...
+    rpc SubscribeAttitudeEuler(SubscribeAttitudeEulerRequest) returns(stream AttitudeEulerResponse) {}
+    rpc SubscribeCameraAttitudeQuaternion(SubscribeCameraAttitudeQuaternionRequest) returns(stream CameraAttitudeQuaternionResponse) {}
+    rpc SubscribeCameraAttitudeEuler(SubscribeCameraAttitudeEulerRequest) returns(stream CameraAttitudeEulerResponse) {}
+    rpc SubscribeGroundSpeedNED(SubscribeGroundSpeedNEDRequest) returns(stream GroundSpeedNEDResponse) {}
+    rpc SubscribeGPSInfo(SubscribeGPSInfoRequest) returns(stream GPSInfoResponse) {}
+    rpc SubscribeBattery(SubscribeBatteryRequest) returns(stream BatteryResponse) {}
+    rpc SubscribeFlightMode(SubscribeFlightModeRequest) returns(stream FlightModeResponse) {}
+    rpc SubscribeHealth(SubscribeHealthRequest) returns(stream HealthResponse) {}
+    rpc SubscribeRCStatus(SubscribeRCStatusRequest) returns(stream RCStatusResponse) {}
 }
```
Comment (shakthi-prashanth-m) [inline telemetry/telemetry.proto]: How do we handle `Unsubscribe()` ? ; we pass `nullptr` in case of  C++.

**Comment by julianoes** [comment]
@JonasVautherin actually what's your opinion on?

`SubscribeRCStatus` vs `SubscribeRcStatus`

**Comment by JonasVautherin** [comment]
@julianoes: I'm okay with both. I just copied "RCStatus" from telemetry_impl.h, but if you find it more readable I can change it to SubscribeRcStatus without a problem.

I see RC the same way as GPS, that we usually don't write as "Gps" (again, I don't have a very strong opinion on that). I have a stronger opinion on "Takeoff" because it is one word (see [Wikipedia](https://en.wikipedia.org/wiki/Takeoff)).

File: telemetry/telemetry.proto (hunk 1 above)
Comment (JonasVautherin) [reply telemetry/telemetry.proto]: That's a good question.

I see two ways:

1. Maybe we can have an `Unsubscribe*` equivalent for each stream, which would make the stream end. I am not sure how clean that is on the server side.
2. (Currently my preferred one) we let the client unsubscribe from the stream whenever it wants, in a pubsub manner.

I checked online on how to make the client unsubscribe from a server stream in gRPC, and [found this for Java](https://github.com/grpc/grpc-java/pull/3115). It allows the client to "cancel" a server stream. I therefore assumed that it should be possible to do that from other languages, but it's right that we will need to test it more thoroughly.

Comment (shakthi-prashanth-m) [reply telemetry/telemetry.proto]: OK Cool.

File: telemetry/telemetry.proto [hunk 2]
```diff
@@ -23,3 +98,68 @@ message Position {
...
+    int32 num_satellites = 1;
+    FixType fix_type = 2;
+}
+
+enum FixType {
+    NO_GPS = 0;
+    NO_FIX = 1;
+    FIX_2D = 2;
+    FIX_3D = 3;
+    FIX_DGPS = 4;
```
Comment (shakthi-prashanth-m) [inline telemetry/telemetry.proto]: Missing **5: RTK float**, **6: RTK fixed**

File: telemetry/telemetry.proto [hunk 3]
```diff
@@ -23,3 +98,68 @@ message Position {
...
+}
+
+message Battery {
+    float voltage = 1;
```
Comment (shakthi-prashanth-m) [inline telemetry/telemetry.proto]: It should be `voltage_v` with units.

File: telemetry/telemetry.proto [hunk 4]
```diff
@@ -23,3 +98,68 @@ message Position {
...
+    MISSION = 3;
+    RETURN_TO_LAUNCH = 4;
+    LAND = 5;
+    OFFBOARD = 6;
+    FOLLOW_ME = 7;
+    UNKNOWN = 1000;
+}
+
+message Health {
+    bool is_gyrometer_calibration_ok = 1;
```
Comment (shakthi-prashanth-m) [inline telemetry/telemetry.proto]: Nice! `is_` is more readable!

File: telemetry/telemetry.proto [hunk 5]
```diff
@@ -23,3 +98,68 @@ message Position {
...
+    bool is_accelerometer_calibration_ok = 2;
+    bool is_magnetometer_calibration_ok = 3;
+    bool is_level_calibration_ok = 4;
+    bool is_local_position_ok = 5;
+    bool is_global_position_ok = 6;
+    bool is_home_position_ok = 7;
+}
+
+message RCStatus {
+    bool was_available_once = 1;
```
Comment (shakthi-prashanth-m) [inline telemetry/telemetry.proto]: This one's Nice too! Its more meaningful to say `was_available_once = true`,

File: telemetry/telemetry.proto [hunk 6]
```diff
@@ -23,3 +98,68 @@ message Position {
...
+    float z = 4;
+}
+
+message EulerAngle {
+    float roll_deg = 1;
+    float pitch_deg = 2;
+    float yaw_deg = 3;
+}
+
+message SpeedNED {
```
[the remark this answers is not in the document]
Comment (JonasVautherin) [reply telemetry/telemetry.proto]: Yeah but both `GroundSpeedNED` and `AirSpeedNED` are "NED" speeds, right? We have a `SpeedNED` in the `GroundSpeedNEDResponse`, but if we had an `AirSpeedNEDResponse`, we could reuse it, couldn't we?

File: telemetry/telemetry.proto (hunk 2 above)
Comment (JonasVautherin) [reply telemetry/telemetry.proto]: Oh, interesting! I copied a comment in the telemetry plugin. Adding that right away!

Comment (JonasVautherin) [reply telemetry/telemetry.proto]: My bad, I misread the comment actually -_-

File: telemetry/telemetry.proto (hunk 3 above)
Comment (JonasVautherin) [reply telemetry/telemetry.proto]: Good catch!

File: telemetry/telemetry.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (shakthi-prashanth-m) [reply telemetry/telemetry.proto]: Yeah, true; we can reuse.

**Comment by julianoes** [comment]
Ok let's leave it as is.

**Pull request merged**